### PR TITLE
http/https proxy

### LINKED ISSUE #7487: HTTP proxy
### Discussed in https://github.com/rustdesk/rustdesk/discussions/5687

<div type='discussions-op-text'>

<sup>Originally posted by **RafaelLinux** August 31, 2022</sup>
In my intranet, we require to use an http(s) proxy. RustDesk nowadays only works with socks proxy, so it shows "Connecting to the RustDesk network" forever and I don't get an ID.

Please, could you add the HTTP(S) proxy feature?

Thank you</div>

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/rustdesk) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/rustdesk/rustdesk/issues/7487">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/rustdesk/rustdesk/issues/7487/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by yuluo-zy**
Add http and https proxy functions, and establish remote proxy server access by filling in the relevant http and https addresses.

for example： “http://192.168.1.3:12345” "https://192.168.1.3:443"

Also supports user authentication。

Fixes #7487 #5687

### CODE REVIEW

**Comment by rustdesk** [comment]
Please use tls-native for macos and windows only, the others use rustls.

image

**Comment by rustdesk** [comment]
@fufesou please review the others.

**Comment by fufesou** [comment]
image

Two bugs here  https://github.com/rustdesk/rustdesk/blob/129548764e28eee6e2921c0c770fb83b7e16621f/libs/hbb_common/src/socket_client.rs#L52
1. Problem with original code. If detecting the address of a proxy, it should detect the address directly instead of using a proxy. 
2. When checking http or https or some other schema the host should be extracted.

**Comment by fufesou** [comment]
Please add tooltip it's default socks5, http or https.

image

**Comment by fufesou** [comment]
I can see http request on rust side is using the proxy. But the request on flutter side does not.

https://github.com/rustdesk/rustdesk/blob/129548764e28eee6e2921c0c770fb83b7e16621f/flutter/lib/models/user_model.dart#L41

TODO: socks5 support on http request.

**Comment by yuluo-zy** [comment]
> 我可以看到 Rust 端的 http 请求正在使用代理。但 flutter 端的请求却没有。
> 
> https://github.com/rustdesk/rustdesk/blob/129548764e28eee6e2921c0c770fb83b7e16621f/flutter/lib/models/user_model.dart#L41
> 
> 通过搜索可以找到flutter端所有的http请求`mainGetApiServer()`
> 
> ====更新====
> 
> 我看不到 rust http 请求在哪里设置了代理。您还应该在 Rust 端设置代理。
> 
> TODO：socks5 支持 http 请求。

Sorry, I only reviewed the code mentioned here https://github.com/rustdesk/rustdesk/discussions/5687#discussioncomment-6996866. For the above content, do I need to add a global http interceptor to complete it, including flutter and rust content?

**Comment by yuluo-zy** [comment]
> image
> 
> Two bugs here
> 
> https://github.com/rustdesk/rustdesk/blob/129548764e28eee6e2921c0c770fb83b7e16621f/libs/hbb_common/src/socket_client.rs#L52
> 
> 1. Bug in the original code. If detecting the address of a proxy, it should detect the address directly instead of using a proxy.
> 2. When checking http or https or some other schema the host should be extracted.

Sorry, I will improve this code

**Comment by yuluo-zy** [comment]
> Please use tls-native for macos and windows only, the others use rustls.
> 
> image
I tested on ubuntu18 and there is no problem with tokio-native-tls. Is rustls used here to support Android and ios? I checked the relevant front-end code rustdesk/flutter/lib/mobile, and there is no proxy setting.

**Comment by fufesou** [comment]
> Sorry, I only reviewed the code mentioned here https://github.com/rustdesk/rustdesk/discussions/5687#discussioncomment-6996866. For the above content, do I need to add a global http interceptor to complete it, including flutter and rust content?

Sounds ok, but please make sure the global http interceptor  only affect current process.

For Socks5, it was implemented very early. There seemed to be no http requests at that time, so it is currently missing. If it is convenient, please add it. If it's inconvenient, forget it.

**Comment by yuluo-zy** [comment]
> test_if_valid_server

Thank you for your reply. Regarding the proxy issue of http requests, I will review the relevant code and propose a specific solution.
But for this PR, it currently seems to have no impact on the access and remote control of the console (except for the test_if_valid_server function above, which I did forget to modify)

> > 抱歉，我只查看了上面提到的代码[#5687（评论）](https://github.com/rustdesk/rustdesk/discussions/5687#discussioncomment-6996866)。对于上面的内容，我是否需要添加一个全局的http拦截器来完成，包括flutter和rust的内容？
> 
> 听起来不错，但请确保全局http拦截器只影响当前进度。
> 
> 对于Socks5来说，很早就实现了。当时好像没有http请求，所以目前是解除的。如果方便的话请补充一下。如果不方便的话，就忘了吧。

**Comment by yuluo-zy** [comment]
> > Sorry, I only reviewed the code mentioned here [#5687 (comment)](https://github.com/rustdesk/rustdesk/discussions/5687#discussioncomment-6996866). For the above content, do I need to add a global http interceptor to complete it, including flutter and rust content?
> 
> Sounds ok, but please make sure the global http interceptor only affect current process.
> 
> For Socks5, it was implemented very early. There seemed to be no http requests at that time, so it is currently missing. If it is convenient, please add it. If it's inconvenient, forget it.

Thank you for your reply. Regarding the proxy issue of http requests, I will review the relevant code and propose a specific solution.
But for this PR, it currently seems to have no impact on the access and remote control of the console (except for the test_if_valid_server function above, which I did forget to modify)

**Comment by rustdesk** [comment]
> > Please use tls-native for macos and windows only, the others use rustls.
> > image
> > I tested on ubuntu18 and there is no problem with tokio-native-tls. Is rustls used here to support Android and ios? I checked the relevant front-end code rustdesk/flutter/lib/mobile, and there is no proxy setting.

Ubuntu 18, Ubuntu 20, 22 has different ssl version, but RustDesk only have one version, so We do not use native tls on Linux.

**Comment by yuluo-zy** [comment]
> I can see http request on rust side is using the proxy. But the request on flutter side does not.
> 
> https://github.com/rustdesk/rustdesk/blob/129548764e28eee6e2921c0c770fb83b7e16621f/flutter/lib/models/user_model.dart#L41
> 
> You can find the all http requests on the flutter side by searching `mainGetApiServer()`
> 
> ==== update ====
> 
> I cann't see where rust http requests have set the proxy. You should also set the proxy on rust side.
> 
> TODO: socks5 support on http request.
For the Rust side code that handles HTTP request proxying, I see that it's implemented using `reqwest` within the `/src/hbbs_http` module. I noticed that within the plugin, `request_plugin_sign` involves HTTP usage. Do these areas need to be modified? Are there any other places that I may have missed?

**Comment by rustdesk** [comment]
Yes, rust side uses reqwest for http, reqwest support socks/http/https proxy very well. But the dart pub http does not support proxy well. I hope we can replace http pub with native reqwest via ffi.

**Comment by yuluo-zy** [comment]
> Yes, rust side uses reqwest for http, reqwest support socks/http/https proxy quite well. But the dart pub http does not support proxy well. I hope we can replace http pub with native reqwest via ffi.

Do you mean that you want me to modify all HTTP requests on the Flutter side to be done through FFI by using Rust's reqwest library?

**Comment by rustdesk** [comment]
Exactly, yes.

**Comment by rustdesk** [comment]
Oh, sorry, you do not have to do this, it is just my expectation since you mentioned. :)

**Comment by rustdesk** [comment]
Thanks, you are right, another PR is better.

File: libs/hbb_common/src/tcp.rs [hunk 1]
```diff
@@ -18,20 +18,22 @@ use tokio::{
...
 #[derive(Clone)]
 pub struct Encrypt(Key, u64, u64);
 
 pub struct FramedStream(
-    Framed<DynTcpStream, BytesCodec>,
-    SocketAddr,
-    Option<Encrypt>,
-    u64,
+    pub(crate) Framed<DynTcpStream, BytesCodec>,
+    pub(crate)SocketAddr,
```
Comment (rustdesk) [inline libs/hbb_common/src/tcp.rs:32]: I guess you did not format files. `pub(crate) Framed` vs `pub(crate)SocketAddr`

File: src/rendezvous_mediator.rs [hunk 2]
```diff
@@ -28,6 +28,8 @@ use hbb_common::{
     udp::FramedSocket,
     AddrMangle, IntoTargetAddr, ResultType, TargetAddr,
 };
+use hbb_common::log::info;
```
Comment (rustdesk) [inline src/rendezvous_mediator.rs]: Please follow our `use` style. (grouped)

File: libs/hbb_common/Cargo.toml [hunk 3]
```diff
@@ -40,10 +40,19 @@ toml = "0.7"
 uuid = { version = "1.3", features = ["v4"] }
 # crash, versions >= 0.29.1 are affected by #GuillaumeGomez/sysinfo/1052
 sysinfo = { git = "https://github.com/rustdesk-org/sysinfo" }
-
+thiserror = "1.0"
+httparse = "1.5"
+base64 = "0.22"
```
Comment (rustdesk) [inline libs/hbb_common/Cargo.toml:45]: `rustdesk/Cargo.toml` has base64 also, could you change to use `hbb_common::base64` in `rustdesk/src`?

File: libs/hbb_common/Cargo.toml [hunk 4]
```diff
@@ -40,10 +40,19 @@ toml = "0.7"
...
+url = "2.2"
 [target.'cfg(not(any(target_os = "android", target_os = "ios")))'.dependencies]
 mac_address = "1.1"
 machine-uid = { git = "https://github.com/21pages/machine-uid" }
+[target.'cfg(not(any(target_os = "macos", target_os = "windows")))'.dependencies]
+tokio-rustls = "0.26"
+webpki-roots = "0.26"
```
Comment (rustdesk) [inline libs/hbb_common/Cargo.toml]: rustls-native-certs is better, or even https://github.com/rustls/rustls-platform-verifier.

Could you please help check if reqwest can use 'rustls-platform-verifier"? Currently, we use rustls-native-certs feature for rustls of reqwest in rustdesk/Cargo.toml.

Comment (rustdesk) [reply libs/hbb_common/Cargo.toml]: webpki-roots also recommend rustls-platform-verifie, https://github.com/rustls/webpki-roots?tab=readme-ov-file#:~:text=rustls%2Dplatform%2Dverifier

**Review by rustdesk — CHANGES_REQUESTED** [review]

File: libs/hbb_common/Cargo.toml (hunk 4 above)
Comment (yuluo-zy) [reply libs/hbb_common/Cargo.toml]: I haven't finished editing yet. Sorry, I forgot to set the draft status. (´ﾟдﾟ`)

Comment (yuluo-zy) [reply libs/hbb_common/Cargo.toml]: Yes, I will make modifications according to the issues you pointed out. (´///☁///`)

**Comment by yuluo-zy** [comment]
> > Yes, rust side uses reqwest for http, reqwest support socks/http/https proxy quite well. But the dart pub http does not support proxy well. I hope we can replace http pub with native reqwest via ffi.
> 
> Do you mean that you want me to modify all HTTP requests on the Flutter side to be done through FFI by using Rust's reqwest library?

https://github.com/rustdesk/rustdesk/blob/1d4c129e9c286f4e2bb866b1629788b6aa2ecabe/src/ui_interface.rs#L712


I noticed that Rust has an existing function for making POST requests. Can I use this function to perform FFI calls? However, it seems that the result would require constant callback queries to check the outcome.

**Comment by rustdesk** [comment]
Yes, you can

**Comment by yuluo-zy** [comment]
> Yes, you can

I encountered a rather awkward problem when testing Flutter's HTTP requests through FFI by calling a Rust HTTP client for verification. I discovered that the domain https://admin.rustdesk.com does not exist. Under these circumstances, how should I proceed with functional testing, especially since the related features require the support of RUSTDESK SERVER PRO?

**Comment by rustdesk** [comment]
@21pages will give you a test server

Please have a good wrapper of the http.post in dart. To make sure minimun impact, please still use http pub when no proxy, and only use native ffi post for proxy.

**Comment by 21pages** [comment]
@yuluo-zy please contact me with email cGFnZXMyMUAxNjMuY29t (base64)

**Comment by rustdesk** [comment]
Good job

**Comment by rustdesk** [comment]
A lot of CI failures, https://github.com/rustdesk/rustdesk/actions/runs/8796163902/job/24138522662

**Comment by yuluo-zy** [comment]
> A lot of CI failures, https://github.com/rustdesk/rustdesk/actions/runs/8796163902/job/24138522662

I'm sorry, I'm fixing these issues today,

**Comment by rustdesk** [comment]
Reverted

**Comment by rustdesk** [comment]
@yuluo-zy ?

**Comment by yuluo-zy** [comment]
Hello, this ci process has been waiting for the application of the mac-arm platform test, the other builds have been passed, do I apply for the merge now, or do I wait for the ci to be fully completed? @rustdesk 

https://github.com/yuluo-zy/rustdesk/actions/runs/8811459353

**Pull request merged**